From 0d92abd3be6d4b91b827030328c3c487e656eea1 Mon Sep 17 00:00:00 2001
From: Ievgen Mutavchi <Ievgen_Mutavchi@comcast.com>
License: GPLV2
Date: Wed, 18 Dec 2019 09:49:19 -0500
Source: COMCAST
Upstream-Status: None
Signed-off-by: Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>
Subject: [PATCH] Make SourceBuffer evictCodedFrames less aggressive

On some streams (with all sync samples and PTS re-odering) evicting
close to current plaback position removes currently playing range
triggering re-buffering.
---
 Source/WebCore/Modules/mediasource/SourceBuffer.cpp | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/Source/WebCore/Modules/mediasource/SourceBuffer.cpp b/Source/WebCore/Modules/mediasource/SourceBuffer.cpp
index 477a966..eb10f59 100644
--- a/Source/WebCore/Modules/mediasource/SourceBuffer.cpp
+++ b/Source/WebCore/Modules/mediasource/SourceBuffer.cpp
@@ -980,13 +980,17 @@ void SourceBuffer::evictCodedFrames(size_t newDataSize)
     MediaTime maximumRangeEnd = currentTime - thirtySeconds;
 
 #if defined(METROLOGICAL)
-    MediaTime microSecond = MediaTime(1, 1000000);
     for (auto& trackBuffer : m_trackBufferMap.values()) {
+        auto t = std::max(MediaTime::zeroTime(), currentTime - MediaTime(2, 1));
         auto prevSync =
-            trackBuffer.samples.decodeOrder().findSyncSamplePriorToPresentationTime(currentTime);
+            trackBuffer.samples.decodeOrder().findSyncSamplePriorToPresentationTime(t);
         if (prevSync != trackBuffer.samples.decodeOrder().rend()) {
             // Don't include the sync frame in the range, just finish right before it.
-            maximumRangeEnd = prevSync->second->presentationTime() - microSecond;
+            prevSync++;
+        }
+        if (prevSync != trackBuffer.samples.decodeOrder().rend()) {
+            maximumRangeEnd = prevSync->second->presentationTime();
+            break;
         }
     }
 #endif
-- 
2.7.4

